go/enclave/rollupchain/rollup_chain.go

```diff
@@ -221,107 +209,115 @@ func (rc *RollupChain) updateState(block *types.Block) (*obscurocore.HeadsAfterL
-	headsAfterParentBlock, err := rc.storage.FetchHeadsAfterL1Block(block.ParentHash())
+	headsAfterParentBlock, err := rc.getHeadsAfterParentBlock(block.ParentHash())
 	if err != nil {
-		if !errors.Is(err, errutil.ErrNotFound) {
-			return nil, fmt.Errorf("could not retrieve parent block state. Cause: %w", err)
-		}
-		// We recursively calculate the state after the parent.
-		parentBlock, err := rc.storage.FetchBlock(block.ParentHash())
-		if err != nil {
-			rc.logger.Crit("Could not retrieve parent block when calculating block state.", log.ErrKey, err)
-		}
-		headsAfterParentBlock, err = rc.updateState(parentBlock)
-		if err != nil {
-			return nil, err
-		}
+		return nil, fmt.Errorf("could not retrieve heads after parent block. Cause: %w", err)
 	}
 
-	if headsAfterParentBlock == nil {
-		rc.logger.Crit(fmt.Sprintf("Could not calculate parent block state when calculating block state. BlockNum=%d. \n Block: %d, Block Parent: %d  ",
-			block.Number(),
-			common.ShortHash(block.Hash()),
-			common.ShortHash(block.Header().ParentHash),
-		))
+	headsAfterL1Block, headRollup, receipts, err := rc.calculateNewHeads(block, headsAfterParentBlock, rollupsInBlock)
+	if err != nil {
+		return nil, fmt.Errorf("could not calculate heads after L1 block. Cause: %w", err)
 	}
 
-	headsAfterL1Block, headRollup, receipts := rc.calculateHeadsAfterL1Block(block, headsAfterParentBlock, rollupsInBlock)
 	rc.logger.Trace(fmt.Sprintf("Calc block state b_%d: Found: %t - r_%d, ",
-		common.ShortHash(block.Hash()),
-		headsAfterL1Block.UpdatedHeadRollup,
-		common.ShortHash(headsAfterL1Block.HeadRollup)))
+		common.ShortHash(block.Hash()), headsAfterL1Block.UpdatedHeadRollup, common.ShortHash(headsAfterL1Block.HeadRollup)))
 
-	var logs []*types.Log
-	for _, receipt := range receipts {
-		logs = append(logs, receipt.Logs...)
-	}
-	err = rc.storage.StoreNewHead(headsAfterL1Block, headRollup, receipts, logs)
+	err = rc.storage.StoreNewHeads(headsAfterL1Block, headRollup, receipts)
 	if err != nil {
-		rc.logger.Crit("Could not store new head.", log.ErrKey, err)
+		return nil, fmt.Errorf("could not store new head. Cause: %w", err)
 	}
 
 	return headsAfterL1Block, nil
 }
 
-func (rc *RollupChain) handleGenesisRollup(b *types.Block, rollups []*obscurocore.Rollup, genesisRollup *obscurocore.Rollup) (genesisState *obscurocore.HeadsAfterL1Block, isGenesis bool) {
-	// the incoming block holds the genesis rollup
-	// calculate and return the new block state
-	// todo change this to an hardcoded hash on testnet/mainnet
-	// TODO - Surface errors instead of (as well as?) returning a bool.
-	if genesisRollup == nil && len(rollups) == 1 {
-		rc.logger.Info("Found genesis rollup", "l1Height", b.NumberU64(), "l1Hash", b.Hash())
+func (rc *RollupChain) handleGenesisRollup(b *types.Block, rollupsInBlock []*obscurocore.Rollup) (isPreGenesis bool, isGenesisRollupInBlock bool, genesisState *obscurocore.HeadsAfterL1Block, err error) {
```

Comment:
I think isPreGenesis should be a known error that can be checked for rather than a boolean.